File: examples/AUI.Example.Views/android/app/src/main/java/com/github/aui/android/MyGLSurfaceView.java
package com.github.aui.android;

import android.content.Context;
import android.opengl.GLSurfaceView;
import android.util.Log;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;

import static android.view.MotionEvent.*;

import java.util.NoSuchElementException;
import java.util.Queue;
import java.util.concurrent.LinkedBlockingDeque;

import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.opengles.GL10;

public class MyGLSurfaceView extends GLSurfaceView implements View.OnTouchListener {
    private static MyGLSurfaceView ourLastSurface;
    private GestureDetector mGestureDetector;

    private final GLSurfaceView.Renderer mRenderer;
    private static Queue<Runnable> ourOnUiThread = new LinkedBlockingDeque<Runnable>();

    public MyGLSurfaceView(Context context) {
        super(context);
        mGestureDetector = new GestureDetector(context, new GestureDetector.OnGestureListener() {
            @Override
            public boolean onDown(MotionEvent motionEvent) {
                return false;
            }

            @Override
            public void onShowPress(MotionEvent motionEvent) {

            }

            @Override
            public boolean onSingleTapUp(MotionEvent motionEvent) {
                return false;
            }

            @Override
            public boolean onScroll(MotionEvent motionEvent, MotionEvent motionEvent1, float velX, float velY) {
                onUiThread(() -> handleScroll((int)motionEvent.getX(), (int)motionEvent.getY(), velX, velY));
                return true;
            }

            @Override
            public void onLongPress(MotionEvent motionEvent) {

            }

            @Override
            public boolean onFling(MotionEvent start, MotionEvent current, float velX, float velY) {
                return false;
            }
        });
        ourLastSurface = this;
        setEGLContextClientVersion(2);
        setEGLConfigChooser(8, 8, 8, 8, 16, 8);
        String internalStoragePath = context.getFilesDir().path();
        setRenderer(mRenderer = new GLSurfaceView.Renderer() {

            @Override
            public void onSurfaceCreated(GL10 gl, EGLConfig config) {
                handleInit(internalStoragePath);
            }

            @Override
            public void onSurfaceChanged(GL10 gl, int width, int height) {
                handleResize(width, height);
            }

            @Override
            public void onDrawFrame(GL10 gl) {
                try {
                    for (Runnable i = null; (i = ourOnUiThread.remove()) != null; ) {
                        i.run();
                    }
                } catch (NoSuchElementException ignored) {}
                handleRedraw();
            }

        });
        setRenderMode(RENDERMODE_CONTINUOUSLY);
        setOnTouchListener(this);
    }

    private static void requestRedraw() {
        ourLastSurface.requestRender();
    }

    public static void onUiThread(Runnable runnable) {
        ourOnUiThread.add(runnable);
        ourLastSurface.requestRender();
    }

    private static native void handleMouseButtonDown(int x, int y);

    private static native void handleMouseButtonUp(int x, int y);
    private static native void handleMouseMove(int x, int y);
    private static native void handleScroll(int originX, int originY, float velX, float velY);

    private static native void handleInit(String internalStoragePath);
    private static native void handleRedraw();
    private static native void handleResize(int width, int height);

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        int x = (int) event.getX();
        int y = (int) event.getY();
        switch (event.getAction()) {
            case ACTION_DOWN:
                onUiThread(() -> handleMouseButtonDown(x, y));
                break;
            case ACTION_MOVE:
                onUiThread(() -> handleMouseMove(x, y));
                break;
            case ACTION_UP:
                onUiThread(() -> handleMouseButtonUp(x, y));
                break;
        }
        mGestureDetector.onTouchEvent(event);
        return true;
    }

}

File: platform/android/lib/src/java/com/github/aui/android/MyGLSurfaceView.kt
// AUI Framework - Declarative UI toolkit for modern C++20
// Copyright (C) 2020-2023 Alex2772, Magicsea
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library. If not, see <http://www.gnu.org/licenses/>.
package com.github.aui.android

import android.content.Context
import android.opengl.GLSurfaceView
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import android.widget.OverScroller
import android.widget.Scroller
import java.util.*
import java.util.concurrent.LinkedBlockingDeque
import javax.microedition.khronos.egl.EGLConfig
import javax.microedition.khronos.opengles.GL10

class MyGLSurfaceView(context: Context) : GLSurfaceView(context), OnTouchListener {
    private val mGestureDetector: GestureDetector
    private var mRenderer: Renderer? = null
    private val mScroller = Scroller(context)

    companion object {
        var ourCurrentSurface: MyGLSurfaceView? = null


        @JvmStatic
        private external fun handleMouseButtonDown(x: Int, y: Int)

        @JvmStatic
        private external fun handleMouseButtonUp(x: Int, y: Int)

        @JvmStatic
        private external fun handleMouseMove(x: Int, y: Int)

        @JvmStatic
        private external fun handleKineticScroll(x: Int, y: Int)

        @JvmStatic
        private external fun handleScroll(originX: Int, originY: Int, velX: Float, velY: Float)

        @JvmStatic
        private external fun handleInit(internalStoragePath: String)

        @JvmStatic
        private external fun handleRedraw()

        @JvmStatic
        private external fun handleResize(width: Int, height: Int)
    }

    init {
        ourCurrentSurface = this
        mGestureDetector = GestureDetector(context, object : GestureDetector.OnGestureListener {
            override fun onDown(motionEvent: MotionEvent): Boolean {
                mScroller.forceFinished(true)
                return true
            }

            override fun onShowPress(motionEvent: MotionEvent) {}
            override fun onSingleTapUp(motionEvent: MotionEvent): Boolean {
                return false
            }

            override fun onScroll(
                motionEvent: MotionEvent,
                motionEvent1: MotionEvent,
                velX: Float,
                velY: Float
            ): Boolean {
                handleScroll(
                    motionEvent.x.toInt(),
                    motionEvent.y.toInt(),
                    velX,
                    velY
                )
                return true
            }

            override fun onLongPress(motionEvent: MotionEvent) {}
            override fun onFling(
                start: MotionEvent,
                current: MotionEvent,
                velX: Float,
                velY: Float
            ): Boolean {
                mScroller.fling(start.x.toInt(), start.y.toInt(),
                                velX.toInt(), velY.toInt(),
                                0, 999999999,
                                0, 999999999)
                return false
            }
        })

        setEGLContextClientVersion(2)
        setEGLConfigChooser(8, 8, 8, 8, 16, 8)
        val internalStoragePath = context.filesDir.path
        setRenderer(object : Renderer {
            override fun onSurfaceCreated(gl: GL10, config: EGLConfig) {
                handleInit(internalStoragePath)
            }

            override fun onSurfaceChanged(gl: GL10, width: Int, height: Int) {
                handleResize(width, height)
            }

            override fun onDrawFrame(gl: GL10) {
                if (mScroller.computeScrollOffset()) {
                    handleKineticScroll(mScroller.currX, mScroller.currY)

                    handleRedraw()
                    requestRender()
                } else {
                    handleRedraw()
                }
            }
        }.also { mRenderer = it })
        renderMode = RENDERMODE_CONTINUOUSLY
        setOnTouchListener(this)
    }

    override fun onTouch(v: View, event: MotionEvent): Boolean {
        val x = event.x.toInt()
        val y = event.y.toInt()
        when (event.action) {
            MotionEvent.ACTION_DOWN -> handleMouseButtonDown(x, y)
            MotionEvent.ACTION_MOVE -> handleMouseMove(x, y)
            MotionEvent.ACTION_UP -> handleMouseButtonUp(x, y)
        }
        mGestureDetector.onTouchEvent(event)
        return true
    }

}